File: src/go/4_BehavioralPatterns/10_Template/templatemethod.go
package templatemethod

import "fmt"

type Game interface {
  Start()
  HaveWinner() bool
  TakeTurn()
  WinningPlayer() int
}

func PlayGame(g Game) {
  g.Start()
  for ;!g.HaveWinner(); {
    g.TakeTurn()
  }
  fmt.Printf("Player %d wins.\n", g.WinningPlayer())
}

type chess struct {
  turn, maxTurns, currentPlayer int
}

func NewGameOfChess() Game {
  return &chess{ 1, 10, 0 }
}

func (c *chess) Start() {
  fmt.Println("Starting a game of chess.")
}

func (c *chess) HaveWinner() bool {
  return c.turn == c.maxTurns
}

func (c *chess) TakeTurn() {
  c.turn++
  fmt.Printf("Turn %d taken by player %d\n",
    c.turn, c.currentPlayer)
  c.currentPlayer = (c.currentPlayer + 1) % 2
}

func (c *chess) WinningPlayer() int {
  return c.currentPlayer
}

func main() {
  chess := NewGameOfChess()
  PlayGame(chess)
}

File: src/js/4_BehavioralPatterns/10_Template/template_method.js
class Game {
  constructor(numberOfPlayers)
  {
    this.numberOfPlayers = numberOfPlayers;
    this.currentPlayer = 0;
  }

  run() {
    this.start();
    while (!this.haveWinner) {
      this.takeTurn();
    }
    console.log(`Player ${this.winningPlayer} wins.`);
  }

  start(){}
  get haveWinner(){}
  takeTurn(){}
  get winningPlayer(){}
}

class Chess extends Game
{
  constructor()
  {
    super(2);
    this.maxTurns = 10;
    this.turn = 1;
  }

  start()
  {
    console.log(
      `Starting a game of chess with ${this.numberOfPlayers} players.`
    );
  }

  get haveWinner()
  {
    return this.turn === this.maxTurns;
  }

  takeTurn() {
    console.log(
      `Turn ${this.turn++} taken by player ${this.currentPlayer}.`
    );
    this.currentPlayer = (this.currentPlayer + 1) % this.numberOfPlayers;
  }

  get winningPlayer()
  {
    return this.currentPlayer;
  }
}

let chess = new Chess();
chess.run();